Start support for PATCH verb.

jjrdk opened the pull request:
I started this work to add support for patch operations as described [in the FHIR specs](https://www.hl7.org/fhir/fhirpatch.html).

This is just an initial draft to get the work started, so any feedback regarding how it should be properly fit into the architecture would be appreciated. Specifically there are some open points about how to integrate with the IAsyncFhirService.

Currently there is support for FHIR patches using Parameters. No json/xml patch is currently supported.

For the different patch operations, there are test cases for different examples, but additional test cases would help make this feature more robust.

jjrdk committed:
Start support for PATCH verb.

jjrdk commented:
[Back link to related issue](https://github.com/FirelyTeam/spark/issues/302)

jjrdk committed:
Add extension methods to create patches from existing resources.

jjrdk commented:
Can someone tell me what is the secret for the integration tests? I have left them unchanged and still they fail :-(

kennethmyhra commented:
@jjrdk will have to look into it

kennethmyhra commented on src/Spark.Engine/Service/AsyncFhirService.cs:
```diff
@@ -318,18 +343,20 @@ public async Task<FhirResponse> HandleInteractionAsync(Entry interaction)
-                    // and "resource not deleted because it doesn't exist" responses, all return NoContent.
-                    // Same with Delete method above.
-                    return Respond.WithCode(HttpStatusCode.NoContent);
+                    return Respond.WithCode(HttpStatusCode.NotFound);
```
@jjrdk This is the line that creates the issue for the integration tests. It needs to be set back to HttpStatusCode.NoContent for them to not fail.

Any reason you set it to HttpStatusCode.NotFound for this PR?

jjrdk replied:
Thanks so much @kennethmyhra for looking at this. I'll revert that change.
So much for boy scout cleanup 🙃

jjrdk committed:
Revert change to FHIR response when object not found.

jjrdk committed:
Merge remote-tracking branch 'remotes/origin/r4/master' into features/patch-verb

kennethmyhra commented on src/Spark.Engine/Extensions/PatchExtensions.cs:
```diff
@@ -0,0 +1,176 @@
+
+    public static class PatchExtensions
+    {
+        private enum Change
```
I think we can probably use the PatchType enum in the Firely FHIR .NET SDK instead: https://github.com/FirelyTeam/firely-net-common/blob/72029833777b73d1fa3d87de783504eaf18e62b8/src/Hl7.Fhir.Support.Poco/Model/PatchParametersExtensions.cs#L10

jjrdk replied:
That enum is private :-(

kennethmyhra commented on src/Spark.Engine/Extensions/PatchExtensions.cs:
```diff
@@ -0,0 +1,176 @@
+    using Hl7.Fhir.Model;
+    using Hl7.Fhir.Serialization;
+
+    public static class PatchExtensions
```
Can we make this internal?

jjrdk replied:
The idea was to have a public way of generating patches from resources. I can make it internal if you prefer.

kennethmyhra commented on src/Spark.Engine/Service/FhirServiceExtensions/IPatchApplicationService.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    using Hl7.Fhir.Model;
+
+    public interface IPatchApplicationService : IFhirServiceExtension
```
Can we rename this to IPatchService

jjrdk replied:
Will do

kennethmyhra commented on src/Spark.Engine/Service/FhirServiceExtensions/PatchApplicationService.cs:
```diff
@@ -0,0 +1,220 @@
+    using Expression = System.Linq.Expressions.Expression;
+    using fhirExpression = Hl7.FhirPath.Expressions;
+
+    public class PatchApplicationService : IPatchApplicationService
```
Also rename to PatchService

jjrdk replied:
Yep

kennethmyhra commented on src/Spark.Engine/Service/AsyncFhirService.cs:
```diff
@@ -296,7 +321,7 @@ private async Task<FhirResponse> CreateAsync(Entry entry)
 
         public FhirResponse HandleInteraction(Entry interaction)
         {
-            return Task.Run(() => HandleInteractionAsync(interaction)).GetAwaiter().GetResult();
```
We need to keep this as Task.Run because of backwards compatibility with ASP.NET FW. I'm looking at other solutions, but for now let's just keep it this way

jjrdk replied:
Got it. Will revert.

kennethmyhra commented on src/Spark.Engine/Service/AsyncFhirService.cs:
```diff
@@ -308,18 +333,23 @@ public async Task<FhirResponse> HandleInteractionAsync(Entry interaction)
-                    var resourceStorage = GetFeature<IResourceStorageService>();
-                    var current = await resourceStorage.GetAsync(interaction.Key.WithoutVersion()).ConfigureAwait(false);
-                    if (current != null && current.IsPresent)
                     {
```
Seems these { have come in by mistake

jjrdk replied:
Will remove

kennethmyhra commented on src/Spark.Engine/Service/AsyncFhirService.cs:
```diff
@@ -308,18 +333,23 @@ public async Task<FhirResponse> HandleInteractionAsync(Entry interaction)
+                        {
+                            return await DeleteAsync(interaction).ConfigureAwait(false);
+                        }
                     }
```
Same here

kennethmyhra commented on src/Spark.Engine/Service/FhirServiceExtensions/PatchApplicationService.cs:
```diff
@@ -0,0 +1,220 @@
+        {
+            foreach (var component in patch.Parameter.Where(x => x.Name == "operation"))
+            {
+                var operationType = component.Part.First(x => x.Name == "type").Value.ToString();
+                var path = component.Part.First(x => x.Name == "path").Value.ToString();
+                var name = component.Part.FirstOrDefault(x => x.Name == "name")?.Value.ToString();
+                var value = component.Part.FirstOrDefault(x => x.Name == "value")?.Value ?? component.Part.FirstOrDefault(x => x.Name == "value")?.Part[0].Value;
+                var valueExpression = System.Linq.Expressions.Expression.Constant(value);
```
Can we refactor this part. Would like to work on a data structure instead of multiple separate variables.

jjrdk replied:
Not sure I can. The individual parts are identified by their name. Since I cannot assume an order of the parts based on the spec, I have to find them by their name.

Do you have a suggestion how improve?

kennethmyhra replied:
My intention was to have it refactored somehow so it would be more easily readable, but I haven't read the Patch-specification deeply enough at the moment to have any good suggestions currently

Let's leave it for now

kennethmyhra commented on src/Spark.Engine.Test/Service/PatchApplicationServiceTests.cs:
```diff
@@ -0,0 +1,384 @@
+            var resource = new Patient {Id = "test", BirthDate = "1930-01-01"};
+            var applier = new PatchApplicationService();
+
+            await Assert
+                .ThrowsAsync<TargetInvocationException>(() => Task.Run(() => applier.Apply(resource, parameters)))
+                .ConfigureAwait(false);
```
Use Throws instead of ThrowsAsync and change the Test method to be synchronous

jjrdk replied:
Changed.

kennethmyhra commented on src/Spark.Engine.Test/Service/PatchApplicationServiceTests.cs:
```diff
@@ -0,0 +1,384 @@
+        [Fact]
+        public void CanApplyPropertyAssignmentPatch()
+        {
+            var xml = @"<Parameters xmlns=""http://hl7.org/fhir"">
+  <parameter>
+    <name value=""operation""/>
+    <part>
+      <name value=""type""/>
+      <valueCode value=""add""/>
+    </part>
+    <part>
+      <name value=""path""/>
+      <valueString value=""Patient""/>
+    </part>
+    <part>
+      <name value=""name""/>
+      <valueString value=""birthDate""/>
+    </part>
+    <part>
+      <name value=""value""/>
+      <valueDate value=""1930-01-01""/>
+    </part>
+  </parameter>
+</Parameters>";
```
Can you put all these in-line resources in the file system. There is a folder called TestData, and under TestData there is a folder for each FHIR Version R3, R4 and so on.

jjrdk replied:
Moved. For good measure, I implemented this feature in R4 because the PATCH verb is not exposed in previous versions.

kennethmyhra commented on src/Spark.Engine/Service/FhirServiceExtensions/PatchApplicationService.cs:
```diff
@@ -0,0 +1,220 @@
+                    new[] { indexExpression.Arguments[0], valueExpression }),
+                _ => result
+            };
+        }
+
+        private static Expression AddValue(Expression result, ConstantExpression value)
+        {
+            return result switch
+            {
+                MemberExpression me when me.Type.IsGenericType
+                                         && GetMethod(me.Type, "Add") != null =>
+                    Expression.Block(
+                        Expression.IfThen(
+                            Expression.Equal(me, Expression.Default(result.Type)),
+                            Expression.Throw(Expression.New(typeof(InvalidOperationException)))),
+                        Expression.Call(me, GetMethod(me.Type, "Add"), value)),
+                MemberExpression me => Expression.Block(
+                    Expression.IfThen(
+                        Expression.NotEqual(me, Expression.Default(result.Type)),
+                        Expression.Throw(Expression.New(typeof(InvalidOperationException)))),
+                    Expression.Assign(me, value)),
+                _ => result
+            };
+        }
+
+        private static Expression DeleteValue(Expression result)
+        {
+            return result switch
+            {
+                IndexExpression indexExpression => Expression.Call(
+                    indexExpression.Object,
+                    GetMethod(indexExpression.Object!.Type, "RemoveAt"),
+                    indexExpression.Arguments),
+                MemberExpression me when me.Type.IsGenericType
+                                         && typeof(List<>).IsAssignableFrom(me.Type.GetGenericTypeDefinition()) =>
+                   Expression.Call(me, GetMethod(me.Type, "Clear")),
+                MemberExpression me => Expression.Assign(me, Expression.Default(me.Type)),
+                _ => result
+            };
+        }
```
Can we move this methods into internal classes, like AddPatchOperation, DeletePatchOperation and so on. And let them all inherit from an internal interface called IPathOperation with a single method Execute.

jjrdk replied:
I don't agree that the command pattern is warranted here. Each operation is pure and I don't see the need to instantiate it one use. If I'm right, then the added benefit that you would get would be to create separate tests for each command. However I think this is just as well achieved by having targeted tests.

Obviously open to discuss this if there is a concern I didn't get.

kennethmyhra replied:
Let's leave this as it is for now

kennethmyhra commented:
Appreciate the effort @jjrdk!

jjrdk committed:
Apply code review changes.

jjrdk committed:
Move test data to separate files in TestData/R4.

kennethmyhra committed:
Remove Task.Run block in async method

kennethmyhra commented:
@jjrdk been looking some more into this and doing some tests.

Trying to run a PATCH via Postman using the Parameters resource below throws the exception: `Expression of type 'Hl7.Fhir.Model.Code' cannot be used for assignment to type 'Hl7.Fhir.Model.Code``1[Hl7.Fhir.Model.MedicationRequest+medicationrequestStatus]'` I think the resource below is valid

Did you consider using FhirPath in place of Linq Expressions? For some of the operations I think this could simplify things

You are maybe already aware that the ToPatch extension method generates delete operations if the property is null. I kind of get why, but this could also be a bit dangerous behavior. So if we in the future going to use this in a client-side manner we would need to be aware of this. In such a scenario we might not always have the full resource

And lastly not sure why the integration tests currently fails, but I have run them locally and they run fine.

```
{
    "resourceType": "Parameters",
    "parameter": [{
            "name": "operation",
            "part": [{
                    "name": "path",
                    "valueString": "MedicationRequest.status"
                }, {
                    "name": "type",
                    "valueCode": "replace"
                }, {
                    "name": "value",
                    "valueCode": "completed"
                }
            ]
        }
    ]
}
````

jjrdk commented:
@kennethmyhra Thanks for getting back on this. I had a look at your test case, and the problem is that the value is parsed by the FHIR parser as a Code (Hl7.Fhir.Model.Code) when the expected value is Code<T> (Hl7.Fhir.Model.Code<T>). These values are unrelated in the type hierarchy, so casting is not an option.

I'll have a closer look at the case before proposing a solution. I'd like to avoid hacking in too many custom conversions.

Regarding FhirPath or LINQ expression, then I actually use the official FhirPath compiler to get the path and then transform this using a visitor. I can have a look at whether it works better to have a visitor that directly updates the resource based on the FhirPath.

On the last point, then I agree that there are going to be a lot of scenarios that have to be considered in terms of access rights and generating patches. I see three ways to approach this:

1. Take away the extension method and not provide the operation
2. Clearly document the behavior and let the user customize the patch afterwards based on the current context.
3. Parameterize the operation to exclude handling specific properties

Maybe a mix of 2 and 3 is preferred.

jjrdk committed:
Handle Code<T> as generic sub type of DataType. Add tests for additional types.

jjrdk committed:
Merge branch 'features/patch-verb' of github.com:jjrdk/spark into features/patch-verb

jjrdk commented:
I had a look at the handling of Code<T> and it's the only generic subtype of DataType, so I implemented a special case for that. I added tests for other common types.

kennethmyhra commented:
@jjrdk Will pull this, just need to have another look at why the integration tests fail

jjrdk commented:
@kennethmyhra Before you pull, let me try to see if the issue is the test data files. I'll try to pull the test data into code and see if I can somehow get this issue fixed.

Would not like to provide broken integration tests.

kennethmyhra commented:
Seems like that worked @jjrdk. That's strange, shouldn't really have had an impact and was not all what the integrations tests complained about.

jjrdk commented:
I'll try to bring them back individually and see if I can pinpoint the offender.

kennethmyhra commented:
Yep, I'll leave the PR until you have done some more investigation

kennethmyhra commented:
@jjrdk The guy who set up the integration tests will have a look at why those are failing for this PR.

Any other suggestion I have is to revert the last  4 commits that made this fail again or pull out the diff of the changes you want to apply and create a new PR and see if that still fails.

andy-a-o commented:
If this may help..

from the html summaries archive:
History001_http---spark-fhir_04-02-21_06-39-55.html

```
FAIL - HI08 - hi08_all_history_whole_system_with_since_test
id HI08
description all history whole system with since
status fail
message Result contains entries in the wrong order.
requires
resourcePatient
methods["create", "update", "delete"]
validates
resource
methods["history-system"]
links
http://hl7.org/fhir/http.html#history
requests
request{:method=>:get, :url=>"http://spark/fhir/_history?_since=2021-04-02T06:38:55Z", :headers=>{"User-Agent"=>"Ruby FHIR Client", "Accept-Charset"=>"utf-8", "Accept"=>"application/fhir+json"}, :path=>"/_history?_since=2021-04-02T06:38:55Z"}
response{:code=>200, :headers=>{"date"=>"Fri, 02 Apr 2021 06:39:55 GMT", "content-type"=>"application/fhir+json; charset=utf-8", "server"=>"Kestrel", "transfer-encoding"=>"chunked"}, :body=>"{\"resourceType\":\"Bundle\",\"id\":\"53e5e4ac-0f73-4414-bf1e-8e2b1089b77d\",\"type\":\"history\",\"total\":121,\"link\":[{\"relation\":\"self\",\"url\":\"http://spark/fhir/_snapshot?id=2f782c44-5b1f-4568-a2cf-8486f0cd9f05&start=0\"},{\"relation\":\"first\",\"url\":\"http://spark/fhir/_snapshot?id=2f782c44-5b1f-4568-a2cf-8486f0cd9f05&start=0\"},{\"relation\":\"last\",\"url\":\"http://spark/fhir/_snapshot?id=2f782c44-5b1f-4568-a2cf-8486f0cd9f05&start=120\"},{\"relation\":\"next\",\"url\":\"http://spark/fhir/_snapshot?id=2f782c44-5b1f-4568-a2cf-8486f0cd9f05&start=20\"}],\"entry\":[{\"request\":{\"method\":\"DELETE\",\"url\":\"http://spark/fhir/Patient/1007/_history/3\"}},{\"fullUrl\":\"http://spark/fhir/Patient/1007/_history/2\",\"resource\":{\"resourceType\":\"Patient\",\"id\":\"1007\",\"meta\":{\"versionId\":\"2\",\"lastUpdated\":\"2021-04-02T06:39:55.087145+00:00\",\"tag\":[{\"system\":\"http://projectcrucible.org\",\"code\":\"testdata\"}]},\"name\":[{\"family\":\"Emerald\",\"given\":[\"Caro\"]}],\"telecom\":[{\"system\":\"email\",\"value\":\"foo@example.com\"}]},\"request\":{\"method\":\"PUT\",\"url\":\"http://spark/fhir/Patient/1007/_history/2\"}},{\"fullUrl\":\"http://spark/fhir/Patient/1007/_history/1\",\"resource\":{\"resourceType\":\"Patient\",\"id\":\"1007\",\"meta\":{\"versionId\":\"1\",\"lastUpdated\":\"2021-04-02T06:39:55.042596+00:00\",\"tag\":[{\"system\":\"http://projectcrucible.org\",\"code\":\"testdata\"}]},\"name\":[{\"family\":\"Emerald\",\"given\":[\"Caro\"]}]},\"request\":{\"method\":\"POST\",\"url\":\"http://spark/fhir/Patient/1007/_history/1\"}},{\"request\":{\"method\":\"DELETE\",\"url\":\"http://spark/fhir/Patient/1006/_history/2\"}},{\"request\":{\"method\":\"DELETE\",\"url\":\"http://spark/fhir/Observation/23/_history/2\"}},{\"request\":{\"method\":\"DELETE\",\"url\":\"http://spark/fhir/Observation/20/_history/2\"}},{\"request\":{\"method\":\"DELETE\",\"url\":\"http://spark/fhir/Condition/6/_history/2\"}},{\"request\":{\"method\":\"DELETE\",\"url\":\"http://spark/fhir/Condition/5/_history/2\"}},{\"request\":{\"method\":\"DELETE\",\"url\":\"http://spark/fhir/Patient/1005/_history/2\"}},{\"fullUrl\":\"http://spark/fhir/Patient/1006/_history/1\",\"resource\":{\"resourceType\":\"Patient\",\"id\":\"1006\",\"meta\":{\"versionId\":\"1\",\"lastUpdated\":\"2021-04-02T06:39:18.214087+00:00\",\"tag\":[{\"system\":\"http://projectcrucible.org\",\"code\":\"testdata\"}]},\"identifier\":[{\"use\":\"official\",\"system\":\"http://projectcrucible.org\",\"value\":\"1617345557\"}],\"name\":[{\"use\":\"official\",\"text\":\"Transaction Crucible\",\"family\":\"Crucible\",\"given\":[\"Transaction\"]}]},\"request\":{\"method\":\"POST\",\"url\":\"http://spark/fhir/Patient/1006/_history/1\"}},{\"request\":{\"method\":\"DELETE\",\"url\":\"http://spark/fhir/Observation/22/_history/2\"}},{\"request\":{\"method\":\"DELETE\",\"url\":\"http://spark/fhir/Observation/21/_history/2\"}},{\"fullUrl\":\"http://spark/fhir/Condition/6/_history/1\",\"resource\":{\"resourceType\":\"Condition\",\"id\":\"6\",\"meta\":{\"versionId\":\"1\",\"lastUpdated\":\"2021-04-02T06:39:18.0805792+00:00\",\"tag\":[{\"system\":\"http://projectcrucible.org\",\"code\":\"testdata\"}]},\"clinicalStatus\":{\"coding\":[{\"system\":\"http://terminology.hl7.org/CodeSystem/condition-clinical\",\"code\":\"resolved\"}]},\"verificationStatus\":{\"coding\":[{\"system\":\"http://terminology.hl7.org/CodeSystem/condition-ver-status\",\"code\":\"refuted\"}]},\"code\":{\"coding\":[{\"system\":\"http://snomed.info/sct\",\"code\":\"414915002\"}]},\"subject\":{\"reference\":\"Patient/1005\"},\"abatementString\":\"Abated at unknown date\"},\"request\":{\"method\":\"POST\",\"url\":\"http://spark/fhir/Condition/6/_history/1\"}},{\"fullUrl\":\"http://spark/fhir/Observation/23/_history/1\",\"resource\":{\"resourceType\":\"Observation\",\"id\":\"23\",\"meta\":{\"versionId\":\"1\",\"lastUpdated\":\"2021-04-02T06:39:18.0652924+00:00\",\"tag\":[{\"system\":\"http://projectcrucible.org\",\"code\":\"testdata\"}]},\"status\":\"final\",\"code\":{\"coding\":[{\"system\":\"http://loinc.org\",\"code\":\"3141-9\"}]},\"subject\":{\"reference\":\"Patient/1005\"},\"valueQuantity\":{\"value\":100.0,\"unit\":\"kg\",\"system\":\"http://unitsofmeasure.org\"}},\"request\":{\"method\":\"POST\",\"url\":\"http://spark/fhir/Observation/23/_history/1\"}},{\"fullUrl\":\"http://spark/fhir/Observation/22/_history/1\",\"resource\":{\"resourceType\":\"Observation\",\"id\":\"22\",\"meta\":{\"versionId\":\"1\",\"lastUpdated\":\"2021-04-02T06:39:17.9974255+00:00\",\"tag\":[{\"system\":\"http://projectcrucible.org\",\"code\":\"testdata\"}]},\"status\":\"final\",\"code\":{\"coding\":[{\"system\":\"http://loinc.org\",\"code\":\"3141-9\"}]},\"subject\":{\"reference\":\"Patient/1005\"},\"valueQuantity\":{\"value\":250.0,\"unit\":\"kg\",\"system\":\"http://unitsofmeasure.org\"}},\"request\":{\"method\":\"POST\",\"url\":\"http://spark/fhir/Observation/22/_history/1\"}},{\"fullUrl\":\"http://spark/fhir/Observation/20/_history/1\",\"resource\":{\"resourceType\":\"Observation\",\"id\":\"20\",\"meta\":{\"versionId\":\"1\",\"lastUpdated\":\"2021-04-02T06:39:17.7610763+00:00\",\"tag\":[{\"system\":\"http://projectcrucible.org\",\"code\":\"testdata\"}]},\"status\":\"final\",\"code\":{\"coding\":[{\"system\":\"http://loinc.org\",\"code\":\"8302-2\"}]},\"subject\":{\"reference\":\"Patient/1005\"},\"valueQuantity\":{\"value\":170.0,\"unit\":\"cm\",\"system\":\"http://unitsofmeasure.org\"}},\"request\":{\"method\":\"POST\",\"url\":\"http://spark/fhir/Observation/20/_history/1\"}},{\"fullUrl\":\"http://spark/fhir/Observation/21/_history/1\",\"resource\":{\"resourceType\":\"Observation\",\"id\":\"21\",\"meta\":{\"versionId\":\"1\",\"lastUpdated\":\"2021-04-02T06:39:17.7611023+00:00\",\"tag\":[{\"system\":\"http://projectcrucible.org\",\"code\":\"testdata\"}]},\"status\":\"final\",\"code\":{\"coding\":[{\"system\":\"http://loinc.org\",\"code\":\"3141-9\"}]},\"subject\":{\"reference\":\"Patient/1005\"},\"valueQuantity\":{\"value\":200.0,\"unit\":\"kg\",\"system\":\"http://unitsofmeasure.org\"}},\"request\":{\"method\":\"POST\",\"url\":\"http://spark/fhir/Observation/21/_history/1\"}},{\"fullUrl\":\"http://spark/fhir/Condition/5/_history/1\",\"resource\":{\"resourceType\":\"Condition\",\"id\":\"5\",\"meta\":{\"versionId\":\"1\",\"lastUpdated\":\"2021-04-02T06:39:17.7611201+00:00\",\"tag\":[{\"system\":\"http://projectcrucible.org\",\"code\":\"testdata\"}]},\"verificationStatus\":{\"coding\":[{\"system\":\"http://terminology.hl7.org/CodeSystem/condition-ver-status\",\"code\":\"confirmed\"}]},\"code\":{\"coding\":[{\"system\":\"http://snomed.info/sct\",\"code\":\"414915002\"}]},\"subject\":{\"reference\":\"Patient/1005\"}},\"request\":{\"method\":\"POST\",\"url\":\"http://spark/fhir/Condition/5/_history/1\"}},{\"fullUrl\":\"http://spark/fhir/Patient/1005/_history/1\",\"resource\":{\"resourceType\":\"Patient\",\"id\":\"1005\",\"meta\":{\"versionId\":\"1\",\"lastUpdated\":\"2021-04-02T06:39:17.7605158+00:00\",\"tag\":[{\"system\":\"http://projectcrucible.org\",\"code\":\"testdata\"}]},\"identifier\":[{\"use\":\"official\",\"system\":\"http://projectcrucible.org\",\"value\":\"1617345557\"}],\"name\":[{\"use\":\"official\",\"text\":\"Transaction Crucible\",\"family\":\"Crucible\",\"given\":[\"Transaction\"]}]},\"request\":{\"method\":\"POST\",\"url\":\"http://spark/fhir/Patient/1005/_history/1\"}},{\"request\":{\"method\":\"DELETE\",\"url\":\"http://spark/fhir/Observation/19/_history/2\"}}]}"}
```

andy-a-o commented:
And this is the test code https://github.com/incendilabs/plan_executor/blob/14a964ae64bfc53d93b794cbfe11c06518b1bde4/lib/tests/suites/history_test.rb#L203

andy-a-o commented:
Problematic records:

```
               "lastUpdated":"2021-04-02T06:39:55.087145+00:00",
               "lastUpdated":"2021-04-02T06:39:55.042596+00:00",
               "lastUpdated":"2021-04-02T06:39:18.214087+00:00",
               "lastUpdated":"2021-04-02T06:39:18.0805792+00:00",
               "lastUpdated":"2021-04-02T06:39:18.0652924+00:00",
               "lastUpdated":"2021-04-02T06:39:17.9974255+00:00",
               "lastUpdated":"2021-04-02T06:39:17.7610763+00:00", <- THIS ONE
               "lastUpdated":"2021-04-02T06:39:17.7611023+00:00",
               "lastUpdated":"2021-04-02T06:39:17.7611201+00:00",
               "lastUpdated":"2021-04-02T06:39:17.7605158+00:00"
```

andy-a-o commented:
Sorry can't tell you more for now, will be able to check it further on the next week, traveling currently.

For now, you could try to run integration tests locally, using the fresh Mongo db and web instance, running via Docker, just as in the Integration Tests workflow setup here, in this repo.

Try running it on the web instance with- and without changes, apply some of them, run again etc. If you can of course.

jjrdk commented:
@andy-a-o Thanks for the investigation. I had found the error message with mis-ordered items in the test errors, but had not dug further. Will take a closer look.

jjrdk committed:
Merge remote-tracking branch 'remotes/origin/r4/master' into features/patch-verb

jjrdk commented:
I tried reverting the latest commits, but that doesn't seem to help :-(

kennethmyhra commented:
I think currently we can rule out that there is anything wrong with this PR. Andy and me will have to investigate the integration tests further

kennethmyhra commented:
Again thanks @jjrdk, much appreciated!
